plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs

```diff
@@ -972,31 +982,97 @@ suggestAddArgument parsedModule Diagnostic {_message, _range}
   where
     message = unifySpaces _message
 
--- TODO use typ to modify type signature
+-- Given a name for the new binding, add a new pattern to the match in the last position,
+-- returning how many patterns are in this match:
+--      addArgToMatch "foo" `bar arg1 arg2 = ...`
+--   => (`bar arg1 arg2 foo = ...`, 2)
+addArgToMatch :: T.Text -> GenLocated l (Match GhcPs body) -> (GenLocated l (Match GhcPs body), Int)
+addArgToMatch name (L locMatch (Match xMatch ctxMatch pats rhs)) =
+  let unqualName = mkRdrUnqual $ mkVarOcc $ T.unpack name
+      newPat = L (noAnnSrcSpanDP1 generatedSrcSpan) $ VarPat NoExtField (noLocA unqualName)
+  in (L locMatch (Match xMatch ctxMatch (pats <> [newPat]) rhs), length pats)
+
+-- Attempt to insert a binding pattern into each match for the given LHsDecl. Also return the declaration's name, and
+-- the number of bound patterns in the declaration's matches, if the HsDecl is a fun bind
+--
+--    insertArg "new_pat" `foo bar baz = 1`
+-- => (`foo bar baz new_pat = 1`, Just ("foo", 2))
+appendFinalPatToMatches :: T.Text -> LHsDecl GhcPs -> TransformT (Either ResponseError) (LHsDecl GhcPs, Maybe (GenLocated SrcSpanAnnN RdrName, Int))
+appendFinalPatToMatches name = \case
+  (L locDecl (ValD xVal (FunBind xFunBind idFunBind mg coreFunBind))) -> do
+    (mg', numPatsMay) <- modifyMgMatchesT' mg (pure . second Just . addArgToMatch name) Nothing combineMatchNumPats
```

Comment:
This all feels a bit roundabout. You've taught two functions to pass up the number of matches, and then changed `modifyMgMatches` to pass a result thorough... but it seems like you could just _look_ for the number of matches independently of the modification, either before or afterwards. Sure, you have to write another function, but I imagine `funBindArgCount` is quite straightforward code and simplifies all the rest.

Reply:
The unfortunate thing is that we don't know which declaration is going to be altered.

Note that we use `modifySmallestDeclWithM` with this function in `addArgumentAction`.

An alternative would be to do two calls to `modifySmallestDeclWithM` (perhaps a variant called `viewSmallestDeclWithM` that just returns some information on the matching decl). The first would do the modification, the second would extract the information (and then the two maybes would be `<|>` together).

I don't personally find that clearer... While I do agree that this code is a little odd, I think it's part and parcel of these sorts of transformations :/

I'm happy to switch it up if you still think it'd be clearer.

Reply:
Okay, so you do quite a bit of finding this decl. You have to go in and find it here to modify it, then you have to go in again to find the matching signature?

Continuing possibly dumb questions: could we not do something like:
```
do
   decl <- findSmallestDecl
   sig <- findSigFor decl
   -- return the modified version in case we want to look at them
   decl' <- modifyDecl decl
   sig' <- modifySig sig
```
Anyway, I won't nitpick more: I'm happy to go with what you think makes sense as a structure here. It just feels like we've got some pieces of work that are currently entangled and I wonder if they could be disentangled.

Reply:
Ah I see your thought. We certainly could do something like this.

However, this not generally how I've seen other AST refactors done with ghc-exactprint. Normally I see modifications done in a lens style `Lens ast Decl`, and I think it makes some sense to continue to do so.

The API you suggest is actually something I've been considering myself for a while, but the exactprint API is far from stable enough to have such niceties (there would probably be a fair amount of CPP involved in the near future...).